---
title: Eccezioni e gestione delle eccezioni - Guida per programmatori C#
ms.custom: seodec18
ms.date: 07/20/2015
helpviewer_keywords:
- exception handling [C#]
- exceptions [C#]
- C# language, exceptions
ms.assetid: 0001887f-4fa2-47e2-8034-2819477e2344
ms.openlocfilehash: c2b991a45a53ce4a8295d6181da11cb09fda6ddb
ms.sourcegitcommit: 986f836f72ef10876878bd6217174e41464c145a
ms.translationtype: HT
ms.contentlocale: it-IT
ms.lasthandoff: 08/19/2019
ms.locfileid: "69590198"
---
# <a name="exceptions-and-exception-handling-c-programming-guide"></a><span data-ttu-id="62c18-102">Eccezioni e gestione delle eccezioni (Guida per programmatori C#)</span><span class="sxs-lookup"><span data-stu-id="62c18-102">Exceptions and Exception Handling (C# Programming Guide)</span></span>
<span data-ttu-id="62c18-103">Le funzionalità di gestione delle eccezioni del linguaggio C# sono utili per gestire qualsiasi situazione imprevista o eccezionale che può verificarsi durante l'esecuzione di un programma.</span><span class="sxs-lookup"><span data-stu-id="62c18-103">The C# language's exception handling features help you deal with any unexpected or exceptional situations that occur when a program is running.</span></span> <span data-ttu-id="62c18-104">Per la gestione delle eccezioni vengono usate le parole chiave `try`, `catch` e `finally` per provare a eseguire azioni che potrebbero non riuscire, per gestire gli errori quando si decide che è ragionevole farlo e per pulire le risorse in un secondo momento.</span><span class="sxs-lookup"><span data-stu-id="62c18-104">Exception handling uses the `try`, `catch`, and `finally` keywords to try actions that may not succeed, to handle failures when you decide that it is reasonable to do so, and to clean up resources afterward.</span></span> <span data-ttu-id="62c18-105">Le eccezioni possono essere generate da CLR (Common Language Runtime), da .NET Framework o qualsiasi libreria di terze parti o dal codice dell'applicazione.</span><span class="sxs-lookup"><span data-stu-id="62c18-105">Exceptions can be generated by the common language runtime (CLR), by the .NET Framework or any third-party libraries, or by application code.</span></span> <span data-ttu-id="62c18-106">Per creare le eccezioni viene usata la parola chiave `throw`.</span><span class="sxs-lookup"><span data-stu-id="62c18-106">Exceptions are created by using the `throw` keyword.</span></span>  
  
 <span data-ttu-id="62c18-107">In molti casi, un'eccezione può essere generata non da un metodo che chiamato direttamente dal codice, ma da qualsiasi metodo più in basso nello stack di chiamate.</span><span class="sxs-lookup"><span data-stu-id="62c18-107">In many cases, an exception may be thrown not by a method that your code has called directly, but by another method further down in the call stack.</span></span> <span data-ttu-id="62c18-108">In questo caso, CLR ripercorrerà lo stack alla ricerca di un metodo con un blocco `catch` per il tipo di eccezione specifico ed eseguirà il primo blocco `catch` trovato.</span><span class="sxs-lookup"><span data-stu-id="62c18-108">When this happens, the CLR will unwind the stack, looking for a method with a `catch` block for the specific exception type, and it will execute the first such `catch` block that if finds.</span></span> <span data-ttu-id="62c18-109">Se non trova un blocco `catch` appropriato nello stack di chiamate, terminerà il processo e visualizzerà un messaggio all'utente.</span><span class="sxs-lookup"><span data-stu-id="62c18-109">If it finds no appropriate `catch` block anywhere in the call stack, it will terminate the process and display a message to the user.</span></span>  
  
 <span data-ttu-id="62c18-110">In questo esempio, un metodo verifica la presenza di divisioni per zero e intercetta l'errore.</span><span class="sxs-lookup"><span data-stu-id="62c18-110">In this example, a method tests for division by zero and catches the error.</span></span> <span data-ttu-id="62c18-111">Senza la gestione delle eccezioni, il programma verrebbe chiuso con un errore **DivideByZeroException non è stata gestita**.</span><span class="sxs-lookup"><span data-stu-id="62c18-111">Without the exception handling, this program would terminate with a **DivideByZeroException was unhandled** error.</span></span>  
  
 [!code-csharp[csProgGuideExceptions#18](~/samples/snippets/csharp/VS_Snippets_VBCSharp/csProgGuideExceptions/CS/Exceptions.cs#18)]  
  
## <a name="exceptions-overview"></a><span data-ttu-id="62c18-112">Panoramica delle eccezioni</span><span class="sxs-lookup"><span data-stu-id="62c18-112">Exceptions Overview</span></span>  
 <span data-ttu-id="62c18-113">Le eccezioni hanno le proprietà seguenti:</span><span class="sxs-lookup"><span data-stu-id="62c18-113">Exceptions have the following properties:</span></span>  
  
- <span data-ttu-id="62c18-114">Le eccezioni sono tipi che derivano fondamentalmente tutti da `System.Exception`.</span><span class="sxs-lookup"><span data-stu-id="62c18-114">Exceptions are types that all ultimately derive from `System.Exception`.</span></span>  
  
- <span data-ttu-id="62c18-115">Racchiudere all'interno di un blocco `try` le istruzioni che potrebbero generare un'eccezione.</span><span class="sxs-lookup"><span data-stu-id="62c18-115">Use a `try` block around the statements that might throw exceptions.</span></span>  
  
- <span data-ttu-id="62c18-116">Quando si verifica un'eccezione nel blocco `try`, il flusso di controllo passa al primo gestore delle eccezioni associato presente in qualsiasi punto nello stack di chiamate.</span><span class="sxs-lookup"><span data-stu-id="62c18-116">Once an exception occurs in the `try` block, the flow of control jumps to the first associated exception handler that is present anywhere in the call stack.</span></span> <span data-ttu-id="62c18-117">In C#, per definire un gestore di eccezioni viene usata la parola chiave `catch`.</span><span class="sxs-lookup"><span data-stu-id="62c18-117">In C#, the `catch` keyword is used to define an exception handler.</span></span>  
  
- <span data-ttu-id="62c18-118">Se non è presente alcun gestore di eccezioni per una determinata eccezione, il programma interrompe l'esecuzione con un messaggio di errore.</span><span class="sxs-lookup"><span data-stu-id="62c18-118">If no exception handler for a given exception is present, the program stops executing with an error message.</span></span>  
  
- <span data-ttu-id="62c18-119">Non intercettare un'eccezione a meno che non sia possibile gestirla e lasciare l'applicazione in uno stato noto.</span><span class="sxs-lookup"><span data-stu-id="62c18-119">Do not catch an exception unless you can handle it and leave the application in a known state.</span></span> <span data-ttu-id="62c18-120">Se si intercetta `System.Exception`, generare di nuovo l'eccezione tramite la parola chiave `throw` alla fine del blocco `catch`.</span><span class="sxs-lookup"><span data-stu-id="62c18-120">If you catch `System.Exception`, rethrow it using the `throw` keyword at the end of the `catch` block.</span></span>  
  
- <span data-ttu-id="62c18-121">Se un blocco `catch` definisce una variabile di eccezione, è possibile usarla per ottenere altre informazioni sul tipo di eccezione che si è verificato.</span><span class="sxs-lookup"><span data-stu-id="62c18-121">If a `catch` block defines an exception variable, you can use it to obtain more information about the type of exception that occurred.</span></span>  
  
- <span data-ttu-id="62c18-122">Le eccezioni possono essere generate in modo esplicito da un programma usando la parola chiave `throw`.</span><span class="sxs-lookup"><span data-stu-id="62c18-122">Exceptions can be explicitly generated by a program by using the `throw` keyword.</span></span>  
  
- <span data-ttu-id="62c18-123">Gli oggetti eccezione contengono informazioni dettagliate sull'errore, ad esempio lo stato dello stack di chiamate e una descrizione testuale dell'errore.</span><span class="sxs-lookup"><span data-stu-id="62c18-123">Exception objects contain detailed information about the error, such as the state of the call stack and a text description of the error.</span></span>  
  
- <span data-ttu-id="62c18-124">Il codice in un blocco `finally` viene eseguito anche se viene generata un'eccezione.</span><span class="sxs-lookup"><span data-stu-id="62c18-124">Code in a `finally` block is executed even if an exception is thrown.</span></span> <span data-ttu-id="62c18-125">Usare un blocco `finally` per rilasciare le risorse, ad esempio per chiudere eventuali flussi o file aperti nel blocco `try`.</span><span class="sxs-lookup"><span data-stu-id="62c18-125">Use a `finally` block to release resources, for example to close any streams or files that were opened in the `try` block.</span></span>  
  
- <span data-ttu-id="62c18-126">Le eccezioni gestite in .NET Framework vengono implementate sulla base del meccanismo di gestione strutturata delle eccezioni in Win32.</span><span class="sxs-lookup"><span data-stu-id="62c18-126">Managed exceptions in the .NET Framework are implemented on top of the Win32 structured exception handling mechanism.</span></span> <span data-ttu-id="62c18-127">Per altre informazioni, vedere [Gestione strutturata delle eccezioni (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) e [A Crash Course on the Depths of Win32 Structured Exception Handling](https://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm) (Corso intensivo su tutti i concetti fondamentali della gestione delle eccezioni strutturata in Win32).</span><span class="sxs-lookup"><span data-stu-id="62c18-127">For more information, see [Structured Exception Handling (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) and [A Crash Course on the Depths of Win32 Structured Exception Handling](https://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm).</span></span>  
  
## <a name="related-sections"></a><span data-ttu-id="62c18-128">Sezioni correlate</span><span class="sxs-lookup"><span data-stu-id="62c18-128">Related Sections</span></span>  
 <span data-ttu-id="62c18-129">Per altre informazioni sulle eccezioni e la gestione delle eccezioni, vedere gli argomenti seguenti:</span><span class="sxs-lookup"><span data-stu-id="62c18-129">See the following topics for more information about exceptions and exception handling:</span></span>  
  
- [<span data-ttu-id="62c18-130">Uso delle eccezioni</span><span class="sxs-lookup"><span data-stu-id="62c18-130">Using Exceptions</span></span>](./using-exceptions.md)  
  
- [<span data-ttu-id="62c18-131">Gestione delle eccezioni</span><span class="sxs-lookup"><span data-stu-id="62c18-131">Exception Handling</span></span>](./exception-handling.md)  
  
- [<span data-ttu-id="62c18-132">Creazione e generazione di eccezioni</span><span class="sxs-lookup"><span data-stu-id="62c18-132">Creating and Throwing Exceptions</span></span>](./creating-and-throwing-exceptions.md)  
  
- [<span data-ttu-id="62c18-133">Eccezioni generate dal compilatore</span><span class="sxs-lookup"><span data-stu-id="62c18-133">Compiler-Generated Exceptions</span></span>](./compiler-generated-exceptions.md)  
  
- [<span data-ttu-id="62c18-134">Procedura: Gestire un'eccezione usando try/catch (Guida per programmatori C#)</span><span class="sxs-lookup"><span data-stu-id="62c18-134">How to: Handle an Exception Using try/catch (C# Programming Guide)</span></span>](./how-to-handle-an-exception-using-try-catch.md)  
  
- [<span data-ttu-id="62c18-135">Procedura: Eseguire codice di pulitura con finally</span><span class="sxs-lookup"><span data-stu-id="62c18-135">How to: Execute Cleanup Code Using finally</span></span>](./how-to-execute-cleanup-code-using-finally.md)  
  
## <a name="c-language-specification"></a><span data-ttu-id="62c18-136">Specifiche del linguaggio C#</span><span class="sxs-lookup"><span data-stu-id="62c18-136">C# Language Specification</span></span>  

<span data-ttu-id="62c18-137">Per altre informazioni, vedere [Eccezioni](~/_csharplang/spec/exceptions.md) nella [Specifica del linguaggio C#](../../language-reference/language-specification/index.md).</span><span class="sxs-lookup"><span data-stu-id="62c18-137">For more information, see [Exceptions](~/_csharplang/spec/exceptions.md) in the [C# Language Specification](../../language-reference/language-specification/index.md).</span></span> <span data-ttu-id="62c18-138">La specifica del linguaggio costituisce il riferimento ufficiale principale per la sintassi e l'uso di C#.</span><span class="sxs-lookup"><span data-stu-id="62c18-138">The language specification is the definitive source for C# syntax and usage.</span></span>
  
## <a name="see-also"></a><span data-ttu-id="62c18-139">Vedere anche</span><span class="sxs-lookup"><span data-stu-id="62c18-139">See also</span></span>

- <xref:System.SystemException>
- [<span data-ttu-id="62c18-140">Guida per programmatori C#</span><span class="sxs-lookup"><span data-stu-id="62c18-140">C# Programming Guide</span></span>](../index.md)
- [<span data-ttu-id="62c18-141">Parole chiave di C#</span><span class="sxs-lookup"><span data-stu-id="62c18-141">C# Keywords</span></span>](../../language-reference/keywords/index.md)
- [<span data-ttu-id="62c18-142">throw</span><span class="sxs-lookup"><span data-stu-id="62c18-142">throw</span></span>](../../language-reference/keywords/throw.md)
- [<span data-ttu-id="62c18-143">try-catch</span><span class="sxs-lookup"><span data-stu-id="62c18-143">try-catch</span></span>](../../language-reference/keywords/try-catch.md)
- [<span data-ttu-id="62c18-144">try-finally</span><span class="sxs-lookup"><span data-stu-id="62c18-144">try-finally</span></span>](../../language-reference/keywords/try-finally.md)
- [<span data-ttu-id="62c18-145">try-catch-finally</span><span class="sxs-lookup"><span data-stu-id="62c18-145">try-catch-finally</span></span>](../../language-reference/keywords/try-catch-finally.md)
- [<span data-ttu-id="62c18-146">Eccezioni</span><span class="sxs-lookup"><span data-stu-id="62c18-146">Exceptions</span></span>](../../../standard/exceptions/index.md)
